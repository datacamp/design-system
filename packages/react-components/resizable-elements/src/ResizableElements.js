import { css } from '@emotion/core';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import CollapseContext from './CollapseContext';

const clip = (value, min, max) => Math.min(Math.max(value, min), max);

const orientationMap = {
  column: {
    adjustableDimension: 'height',
    adjustablePosition: 'top',
    clientPosition: 'clientY',
    fixedDimension: 'width',
    offset: 'y',
  },
  row: {
    adjustableDimension: 'width',
    adjustablePosition: 'left',
    clientPosition: 'clientX',
    fixedDimension: 'height',
    offset: 'x',
  },
};

class ResizableElements extends PureComponent {
  static propTypes = {
    children: PropTypes.node,
    minSize: PropTypes.number,
    orientation: PropTypes.oneOf(['row', 'column']),
    proportions: PropTypes.arrayOf(PropTypes.number),
  };

  static defaultProps = {
    children: null,
    minSize: 30,
    orientation: 'row',
    proportions: null,
  };

  state = {
    collapsedFirstElement: false,
    collapsedLastElement: false,
    draggingDiff: 0,
    draggingIndex: -1,
    startDraggingAt: null,
  };

  componentWillMount() {
    const { children, proportions } = this.props;
    const nbChildren = this.getNumberOfElements();
    const childWidth = 100 / nbChildren;

    this.setState({
      // TODO: Need to split in sizesPerc and sizesPx to remove calculations from render function.
      //       This will also make auto-uncollapsing trivial.
      sizes:
        proportions === null
          ? React.Children.map(children, () => childWidth)
          : proportions,
    });
  }

  getNumberOfElements() {
    const { children } = this.props;
    return React.Children.count(children);
  }

  parentRef = React.createRef();

  startDragging = (index) => (event) => {
    const { sizes } = this.state;
    const { minSize, orientation } = this.props;
    const orientationConf = orientationMap[orientation];
    const parentBox = this.parentRef.current.getBoundingClientRect();
    const parentSize = parentBox[orientationConf.adjustableDimension];
    const parentOffset = parentBox[orientationConf.offset];
    let draggingMin = (-sizes[index] / 100) * parentSize;
    let draggingMax = sizes[index + 1]
      ? (sizes[index + 1] / 100) * parentSize
      : event[orientationConf.clientPosition] - (parentOffset + parentSize);

    draggingMin += minSize;
    draggingMax -= minSize;

    // state is only used in later setState callbacks, and causes linter warnings due to https://github.com/yannickcr/eslint-plugin-react/issues/1697
    this.setState({
      draggingDiff: 0,
      draggingIndex: index,
      draggingMax,
      draggingMin,
      startDraggingAt: event[orientationConf.clientPosition],
    });
  };

  drag = (event) => {
    // Get it out the event, to persist event property
    const { orientation } = this.props;
    const orientationConf = orientationMap[orientation];
    const clientLoc = event[orientationConf.clientPosition];
    this.setState(({ draggingMax, draggingMin, startDraggingAt }) => ({
      draggingDiff: clip(clientLoc - startDraggingAt, draggingMin, draggingMax),
    }));
  };

  stopDragging = () => {
    this.setState(
      (
        {
          collapsedFirstElement,
          collapsedLastElement,
          draggingDiff,
          draggingIndex,
          sizes,
        },
        { minSize, orientation },
      ) => {
        const nbElements = this.getNumberOfElements();
        const orientationConf = orientationMap[orientation];
        const parentBox = this.parentRef.current.getBoundingClientRect();
        const parentSize = parentBox[orientationConf.adjustableDimension];
        const newSizes = [...sizes];
        const diff = (draggingDiff / parentSize) * 100;
        const minProportion = (minSize / parentSize) * 100;
        if (draggingIndex === 0 && collapsedFirstElement) {
          // uncollapse first element
          newSizes[0] = minProportion + diff;
          newSizes[1] = sizes[1] + sizes[0] - newSizes[0];
        } else if (draggingIndex === nbElements - 2 && collapsedLastElement) {
          // uncollapse last element
          newSizes[draggingIndex + 1] = minProportion - diff;
          newSizes[draggingIndex] =
            sizes[draggingIndex] +
            sizes[draggingIndex + 1] -
            newSizes[draggingIndex + 1];
        } else {
          newSizes[draggingIndex] = sizes[draggingIndex] + diff;
          newSizes[draggingIndex + 1] = sizes[draggingIndex + 1] - diff;
        }

        return {
          collapsedFirstElement:
            draggingIndex === 0
              ? newSizes[0] <= minProportion
              : collapsedFirstElement,
          collapsedLastElement:
            draggingIndex === nbElements - 2
              ? newSizes[nbElements - 1] <= minProportion
              : collapsedLastElement,
          draggingDiff: 0,
          draggingIndex: -1,
          draggingMax: null,
          draggingMin: null,
          sizes: newSizes,
          startDraggingAt: null,
        };
      },
    );
  };

  toggleFirstElement = () => {
    this.setState(({ collapsedFirstElement }) => ({
      collapsedFirstElement: !collapsedFirstElement,
    }));
  };

  toggleLastElement = () => {
    this.setState(({ collapsedLastElement }) => ({
      collapsedLastElement: !collapsedLastElement,
    }));
  };

  render() {
    const { collapsedFirstElement, collapsedLastElement, sizes } = this.state;
    const { children, minSize, orientation } = this.props;
    const orientationConf = orientationMap[orientation];
    const nbChildren = this.getNumberOfElements();
    const splitterCorrection = ((nbChildren - 1) * 5) / nbChildren;

    return (
      <div
        css={css({
          height: '100%',
          position: 'relative',
          width: '100%',
        })}
        ref={this.parentRef}
      >
        <CollapseContext.Provider
          value={{
            collapsedFirstElement,
            collapsedLastElement,
            toggleFirstElement: this.toggleFirstElement,
            toggleLastElement: this.toggleLastElement,
          }}
        >
          {this.state.draggingIndex !== -1 && (
            <div
              css={css({
                background: 'transparent',
                height: '100%',
                left: 0,
                position: 'fixed',
                top: 0,
                width: '100%',
                zIndex: 100,
              })}
              onMouseMove={this.drag}
              onMouseUp={this.stopDragging}
              role="separator"
            />
          )}
          {React.Children.map(children, (child, index) => {
            let sizePerc = 0;
            let sizePx = 0;
            const collapsed =
              (index === 0 && collapsedFirstElement) ||
              (index === nbChildren - 1 && collapsedLastElement);
            if (collapsed) {
              sizePx = minSize;
            } else {
              sizePerc = sizes[index];
              if (index === 1 && collapsedFirstElement) {
                sizePerc += sizes[index - 1];
                sizePx -= minSize;
              }
              if (index === nbChildren - 2 && collapsedLastElement) {
                sizePerc += sizes[index + 1];
                sizePx -= minSize;
              }
            }
            return (
              <React.Fragment>
                <div
                  css={{
                    float: 'left',
                    overflow: 'hidden',
                    [orientationConf.fixedDimension]: '100%',
                    position: 'relative',
                  }}
                  style={{
                    [orientationConf.adjustableDimension]: `calc(${sizePerc}% + ${
                      sizePx - splitterCorrection
                    }px)`,
                  }}
                >
                  {child}
                </div>
                {index < nbChildren - 1 && (
                  <div
                    css={css(
                      orientation === 'column' && {
                        cursor: 'row-resize',
                        height: 5,
                        width: '100%',
                      },
                      orientation === 'row' && {
                        cursor: 'col-resize',
                        height: '100%',
                        width: 5,
                      },
                      this.state.draggingIndex === index && {
                        backgroundColor: '#e5e1da',
                      },
                      { ':hover': { backgroundColor: '#e5e1da' } },
                    )}
                    onMouseDown={this.startDragging(index)}
                    role="separator"
                    style={{
                      [orientationConf.adjustablePosition]:
                        this.state.draggingIndex === index
                          ? `${this.state.draggingDiff}px`
                          : '0px',
                    }}
                  />
                )}
              </React.Fragment>
            );
          })}
        </CollapseContext.Provider>
      </div>
    );
  }
}

export default ResizableElements;
