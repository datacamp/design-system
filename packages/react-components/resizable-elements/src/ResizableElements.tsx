import { css } from '@emotion/react';
import React, { useCallback } from 'react';

import calculateSizes from './calculateSizes';
import CollapseContext from './CollapseContext';
import reducer from './reducer';

const orientationMap = {
  column: {
    adjustableDimension: 'height',
    adjustablePosition: 'top',
    clientPosition: 'clientY',
    fixedDimension: 'width',
    offset: 'y',
  },
  row: {
    adjustableDimension: 'width',
    adjustablePosition: 'left',
    clientPosition: 'clientX',
    fixedDimension: 'height',
    offset: 'x',
  },
} as const;

interface ResizableElementsProps {
  /**
   * The elements to make "resizable". They will each be rendered inside a div that can change in size.
   */
  children: React.ReactNode[];
  /**
   * The size of a "collapsed" element in pixels.
   */
  collapsedSize?: number;
  /**
   * An array of percentage values. This should be the same length as the number of elements.
   * This determines the initial sizes of the elements. When not provided the elements will
   * default to equal sizes.
   */
  initialProportions?: number[];
  /**
   * The layout of the child elements.
   */
  orientation: 'row' | 'column';
}

const ResizableElements = ({
  children,
  collapsedSize = 30,
  initialProportions,
  orientation = 'row',
}: ResizableElementsProps): React.ReactElement => {
  const nbChildren = React.Children.count(children);

  const [
    {
      collapsedFirstElement,
      collapsedLastElement,
      draggingState,
      sizePercentages,
    },
    dispatch,
  ] = React.useReducer(reducer, {
    collapsedFirstElement: false,
    collapsedLastElement: false,
    collapsedSize,
    sizePercentages:
      initialProportions ||
      [...Array(nbChildren).keys()].map(() => 100 / nbChildren),
  });

  const containerRef = React.useRef<HTMLDivElement>(null);

  const {
    adjustableDimension,
    adjustablePosition,
    clientPosition,
    fixedDimension,
    offset,
  } = orientationMap[orientation];

  const startDragging = useCallback(
    (index: number) => (
      event: React.MouseEvent<HTMLDivElement, MouseEvent>,
    ): void => {
      if (!containerRef.current) return;
      const parentBox = containerRef.current.getBoundingClientRect();
      const parentSize = parentBox[adjustableDimension];
      const parentOffset = parentBox[offset];
      const startPosition = event[clientPosition];

      dispatch({
        index,
        parentOffset,
        parentSize,
        startPosition,
        type: 'startDragging',
      });
    },
    [adjustableDimension, clientPosition, offset],
  );

  const drag = useCallback(
    (event: React.MouseEvent<HTMLDivElement, MouseEvent>): void => {
      const clientLoc = event[clientPosition];
      dispatch({ clientLoc, type: 'dragUpdate' });
    },
    [clientPosition],
  );

  const stopDragging = useCallback((): void => {
    if (!containerRef.current) return;

    const parentBox = containerRef.current.getBoundingClientRect();
    const parentSize = parentBox[adjustableDimension];

    dispatch({ parentSize, type: 'stopDragging' });
  }, [adjustableDimension]);

  const toggleFirstElement = useCallback((): void => {
    dispatch({ type: 'toggleFirstElement' });
  }, []);

  const toggleLastElement = useCallback((): void => {
    dispatch({ type: 'toggleLastElement' });
  }, []);

  const sizes = calculateSizes(
    sizePercentages,
    collapsedFirstElement,
    collapsedLastElement,
    collapsedSize,
  );

  return (
    <div
      css={css({
        display: 'flex',
        flexDirection: orientation,
        height: '100%',
        position: 'relative',
        width: '100%',
      })}
      ref={containerRef}
    >
      {draggingState && (
        <div
          css={css({
            background: 'transparent',
            height: '100%',
            left: 0,
            position: 'fixed',
            top: 0,
            width: '100%',
            zIndex: 100,
          })}
          onMouseMove={drag}
          onMouseUp={stopDragging}
          role="separator"
        />
      )}
      {React.Children.map(children, (child, index) => {
        return (
          <CollapseContext.Provider
            value={{
              collapsedFirstElement,
              collapsedLastElement,
              toggleFirstElement,
              toggleLastElement,
            }}
          >
            <div
              css={{
                [fixedDimension]: '100%',
                float: 'left',
                position: 'relative',
              }}
              style={{
                [adjustableDimension]: sizes[index],
              }}
            >
              {child}
            </div>
            {index < nbChildren - 1 && (
              <div
                css={css(
                  orientation === 'column' && {
                    cursor: 'row-resize',
                    height: 5,
                    width: '100%',
                  },
                  orientation === 'row' && {
                    cursor: 'col-resize',
                    height: '100%',
                    width: 5,
                  },
                  draggingState &&
                    draggingState.draggingIndex === index && {
                      backgroundColor: '#e5e1da',
                    },
                  { ':hover': { backgroundColor: '#e5e1da' } },
                )}
                onMouseDown={startDragging(index)}
                role="separator"
                style={{
                  [adjustablePosition]:
                    draggingState && draggingState.draggingIndex === index
                      ? `${draggingState.draggingDiff}px`
                      : '0px',
                }}
              />
            )}
          </CollapseContext.Provider>
        );
      })}
    </div>
  );
};

export default ResizableElements;
