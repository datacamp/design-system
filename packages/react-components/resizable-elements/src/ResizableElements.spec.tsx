import { fireEvent, render } from '@testing-library/react';
import React from 'react';

import ResizableElements from './ResizableElements';

describe('ResizableElements', () => {
  it('renders the correct sized elements', () => {
    const { container } = render(
      <ResizableElements orientation="row">
        <div>firstChild</div>
        <div>secondChild</div>
        <div>thirdChild</div>
      </ResizableElements>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('eventHandlers', () => {
    let useReducerSpy: jest.SpyInstance;

    beforeEach(() => {
      useReducerSpy = jest.spyOn(React, 'useReducer');
    });

    it('dispatches a startDragging action', () => {
      const dispatch = jest.fn();
      useReducerSpy.mockReturnValue([
        {
          collapsedFirstElement: false,
          collapsedLastElement: false,
          draggingState: null,
          sizePercentages: [33, 33, 33],
        },
        dispatch,
      ]);

      const { getAllByRole } = render(
        <ResizableElements orientation="row">
          <div>firstChild</div>
          <div>secondChild</div>
          <div>thirdChild</div>
        </ResizableElements>,
      );

      const separators = getAllByRole('separator');

      fireEvent.mouseDown(separators[0]);

      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledWith(
        expect.objectContaining({ type: 'startDragging' }),
      );
    });

    it('dispatches a dragUpdate action', () => {
      const dispatch = jest.fn();
      useReducerSpy.mockReturnValue([
        {
          collapsedFirstElement: false,
          collapsedLastElement: false,
          draggingState: {},
          sizePercentages: [33, 33, 33],
        },
        dispatch,
      ]);

      const { getAllByRole } = render(
        <ResizableElements orientation="row">
          <div>firstChild</div>
          <div>secondChild</div>
          <div>thirdChild</div>
        </ResizableElements>,
      );

      const separators = getAllByRole('separator');

      fireEvent.mouseMove(separators[0]);

      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledWith(
        expect.objectContaining({ type: 'dragUpdate' }),
      );
    });

    it('dispatches a stopDraggin action', () => {
      const dispatch = jest.fn();
      useReducerSpy.mockReturnValue([
        {
          collapsedFirstElement: false,
          collapsedLastElement: false,
          draggingState: {},
          sizePercentages: [33, 33, 33],
        },
        dispatch,
      ]);

      const { getAllByRole } = render(
        <ResizableElements orientation="row">
          <div>firstChild</div>
          <div>secondChild</div>
          <div>thirdChild</div>
        </ResizableElements>,
      );

      const separators = getAllByRole('separator');

      fireEvent.mouseUp(separators[0]);

      expect(dispatch).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledWith(
        expect.objectContaining({ type: 'stopDragging' }),
      );
    });
  });
});
