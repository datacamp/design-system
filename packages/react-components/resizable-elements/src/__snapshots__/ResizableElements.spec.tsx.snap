// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResizableElements renders the correct sized elements 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 100%;
  position: relative;
  width: 100%;
}

.emotion-1 {
  height: 100%;
  float: left;
  position: relative;
}

.emotion-2 {
  cursor: col-resize;
  height: 100%;
  width: 5px;
}

.emotion-2:hover {
  background-color: #e5e1da;
}

<div
  class="emotion-0"
>
  <div
    class="emotion-1"
    style="width: calc(33.333333333333336% + -3.3333333333333335px);"
  >
    <div>
      firstChild
    </div>
  </div>
  <div
    class="emotion-2"
    role="separator"
    style="left: 0px;"
  />
  <div
    class="emotion-1"
    style="width: calc(33.333333333333336% + -3.3333333333333335px);"
  >
    <div>
      secondChild
    </div>
  </div>
  <div
    class="emotion-2"
    role="separator"
    style="left: 0px;"
  />
  <div
    class="emotion-1"
    style="width: calc(33.333333333333336% + -3.3333333333333335px);"
  >
    <div>
      thirdChild
    </div>
  </div>
</div>
`;
